File: Language/C#/Book/tmp/LINQ实战/Chapter13.VB - LinqBooks/LinqBooks.Web/CustomQueryOperators.vb
Imports System.Runtime.CompilerServices
Imports LinqBooks.Entities

Public Module CustomQueryOperators

  <Extension()> _
  Public Function MaxElement(Of TElement, TData As IComparable(Of TData))( _
    ByVal source As IEnumerable(Of TElement), _
    ByVal selector As Func(Of TElement, TData)) As TElement

    If source Is Nothing Then
      Throw New ArgumentNullException("source")
    End If
    If selector Is Nothing Then
      Throw New ArgumentNullException("selector")
    End If

    Dim firstElement As Boolean = True
    Dim result As TElement ' = default(TElement)
    Dim maxValue As TData ' = default(TData)
    For Each element In source
      Dim candidate = selector(element)
      If firstElement Or (CType(candidate, IComparable).CompareTo(maxValue) > 0) Then
        firstElement = False
        maxValue = candidate
        result = element
      End If
    Next
    Return result
  End Function

  <Extension()> _
  Public Function TotalPrice(ByVal books As IEnumerable(Of Book)) As Decimal
    If books Is Nothing Then
      Throw New ArgumentNullException("books")
    End If

    Dim result As Decimal = 0
    For Each book As Book In books
      If Not book Is Nothing AndAlso book.Price.HasValue Then
        result += book.Price.Value
      End If
    Next
    Return result
  End Function
End Module

File: Language/C#/Book/LINQ实战/第13章 应用程序各层中的LINQ/Chapter13.CSharp - LinqBooks/LinqBooks.Web/CustomQueryOperators.cs
#region Imports

using System;
using System.Collections.Generic;
using System.Linq;

using LinqBooks.Entities;

#endregion Imports

namespace LinqBooks.Web
{
  public static class CustomQueryOperators
  {
    public static TElement MaxElement<TElement, TData>(
      this IEnumerable<TElement> source,
      Func<TElement, TData> selector)
      where TData : IComparable<TData>
    {
      if (source == null)
        throw new ArgumentNullException("source");
      if (selector == null)
        throw new ArgumentNullException("selector");

      TElement result = default(TElement);
      TData maxValue = default(TData);
      foreach (TElement element in source)
      {
        var candidate = selector(element);
        if ((maxValue == null) || ((IComparable)candidate).CompareTo(maxValue) > 0)
        {
          maxValue = candidate;
          result = element;
        }
      }
      return result;
    }

    public static Decimal TotalPrice(this IEnumerable<Book> books)
    {
      if (books == null)
        throw new ArgumentNullException("books");

      Decimal result = 0;
      foreach (Book book in books)
        if ((book != null) && (book.Price.HasValue))
          result += book.Price.Value;
      return result;
    }
  }
}